perfkitbenchmarker/providers/docker/docker_virtual_machine.py

```diff
@@ -0,0 +1,303 @@
+    """
+    Sets the internal and external IP address for the Container.
+    """
+    info, returnCode = self._GetContainerInfo()
+    ip = False
+
+    if len(info) > 0 and returnCode == 0:
+      ip = info[0]['NetworkSettings']['IPAddress'].encode('ascii')
+      logging.info("IP: " + str(ip))
+      self.ip_address = ip
+      self.internal_ip = ip
+    else:
+      logging.warning("IP address information not found")
+
+  def _RemoveIfExists(self):
+    if self._Exists():
+      self._Delete()
+
+  def _GetContainerInfo(self):
+    """
+    Gets Container information from Docker Inspect. Returns the information,
+    if there is any and a return code. 0
+    """
+    logging.info("Checking Container Information")
+    inspect_cmd = ['docker', 'inspect', self.name]
+    info, _, returnCode = vm_util.IssueCommand(inspect_cmd, suppress_warning=True)
+    info = json.loads(info)
+    return info, returnCode
+
+  def _ConfigureProxy(self):
+    """
+    In Docker containers environment variables from /etc/environment
+    are not sourced - this results in connection problems when running
+    behind proxy. Prepending proxy environment variables to bashrc
+    solves the problem. Note: APPENDING to bashrc will not work because
+    the script exits when it is NOT executed in interactive shell.
+    """
+    if FLAGS.http_proxy:
+      http_proxy = "sed -i '1i export http_proxy=%s' /etc/bash.bashrc"
+      self.RemoteCommand(http_proxy % FLAGS.http_proxy)
+    if FLAGS.https_proxy:
+      https_proxy = "sed -i '1i export https_proxy=%s' /etc/bash.bashrc"
+      self.RemoteCommand(https_proxy % FLAGS.http_proxy)
+    if FLAGS.ftp_proxy:
+      ftp_proxy = "sed -i '1i export ftp_proxy=%s' /etc/bash.bashrc"
+      self.RemoteCommand(ftp_proxy % FLAGS.ftp_proxy)
+
+  def _BuildVolumesBody(self):
+    """
+    Constructs volumes-related part of create command for Docker Container
+    """
+    volumes = []
+
+    for scratch_disk in self.scratch_disks:
+      vol_string = scratch_disk.volume_name + ":" + scratch_disk.mount_point
+      volumes.append('-v')
+      volumes.append(vol_string)
+
+    return volumes
+
+  def _LocalImageExists(self, docker_image_name):
+    """
+    Checks if an image exists locally
+    Returns boolean
+    """
+    inspect_cmd = ['docker', 'image', 'inspect', docker_image_name]
+    info, _, returnCode = vm_util.IssueCommand(inspect_cmd, suppress_warning=True)
+    info = json.loads(info)
+    logging.info("Checking if Docker Image Exists")
+    if len(info) > 0 and returnCode == 0:
+      logging.info("Image exists")
+      return True
+    logging.info("Image does not exist")
+    return False
+
+  def _BuildImageLocally(self):
+    """
+    Build Container Image Locally
+    """
+
+    # Dockerfiles located at PerfKitBenchmarker/docker/<containerImage>/Dockerfile
+    directory = os.path.dirname(
+        data.ResourcePath(os.path.join('docker', self.container_image, 'Dockerfile')))
+    build_cmd = [
```

Comment:
Would be nice if these build functions could be shared with https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/container_service.py#L321

Reply:
I agree that using those build functions would be nice, and I looked into it when I first wrote this. But these are part of BaseContainerRegistry, and the Docker provider type doesn't really have a registry since the image it uses is just built locally.